### DISCUSSION TITLE
Quirks when setting Hostname and Topic

### DISCUSSION THREAD

**Discussion by g0lgs** [arendst/Tasmota#15995] [Q&A] [answered]
Hi,

I am writing a script (in Perl as it happens) that I will use to program a significant number of devices 
(over the USB / Serial) that will use a ESP12F module.

I have discovered that if I set the Topic like:

Topic tasmota/test
or 
Topic test
or
Topic tasmota/%hostname%

Without having set the hostname (which I was intending to do later when the devices get to the final location) then the message when it connects to the configured Wifi the hostname it uses is no longer the default tasmota-%06-%04 format.

So 

WIF: Connecting to AP1 XXXXXXXX in mode 11n as tasmota-B13C96-7318..

Becomes:

WIF: Connecting to AP1 XXXXXXXX in mode 11n as tasmota-test-7318..
or
WIF: Connecting to AP1 XXXXXXXX in mode 11n as test-7318..
or
WIF: Connecting to AP1 XXXXXXXX in mode 11n as tasmota-hostname-7318..

(i.e. It changes depending on what is entered as the Topic value)

If you set the Hostname to something other than the default %s-%04d

Then the message is exactly as I would expect.

It really does appear to be using the hostname shown when it connects to the WiFi too (or so my router reports).

I have tested this with Tasmota V11.1.0 and V12.0.2

**Comment by sfromis** [comment, marked as answer]
Using `%hostname%` that way is not officially supported, and may not work in all contexts. If you are doing scripts to config, no need for such tricks, just config all the values you want, and no risk of it not working after all, in some situation.

**Reply by g0lgs** [reply]
I get that, but it still does odd things even without using %hostname%

**Reply by ascillato2** [reply]
The default HOSTNAME is made using the TOPIC, so making `Topic tasmota/%hostname%` won't work as it is not supported.

**Reply by ascillato2** [reply]
The line code in charge of that is:

https://github.com/arendst/Tasmota/blob/cc22151ca9869e06b62aa4ccfe709d42d21b54ff/tasmota/include/tasmota.h#L151

That means that HOSTNAME is TOPIC-XXXX where XXXX are the last 4 digits of the MAC of that device. As the default name depends on the topic's name, your command won't work.

If you want a different hostname, set it using HOSTNAME command. If you want a different topic, set it using TOPIC command.

**Reply by g0lgs** [reply]
You appear to be missing the point.

If I set the Topic without setting the hostname, no matter what I set it too the Hostname is affected unless that has already been set.

I am happy to set both in my script (now that I have discovered this seemingly odd behaviour).

**Reply by ascillato** [reply]
I understand what you want to do, but please, take a look to the code I post. The HOSTAME is by default the TOPIC + 4numbers. So, if you change the topic, the HOSTNAME will change too because in the default, they are **already linked**.

For example:

If you set the TOPIC to be equals to MYTOPIC, the Hostname will be MYTOPIC-1324
So, you can't set the reverse: "topic depending on the Hostname" because by default is the other way round.

What you can do is to override that default behavior (Hostname depending on the topic) by setting a Hostname yourself. Then, no matter what topic you set, the Hostname will remain exactly to the one you entered.

**Reply by ascillato** [reply]
What you have experienced is the designed behavior so as to have all devices with different topic and different hostname when they are with the default parameters (like recently flashed devices). This was designed to prevent network collisions if you flash and connect several Tasmotas at the same time but without configuring them. In fact, configuring topic and hostname are only for the preference of the user. Those changes are not really needed for using Tasmota since all devices will have different values by default.

The hostname is TOPIC-Last4MacNumbers, and the default topic is TASMOTA-Last6MacNumbers. So with that, the default HOSTNAME will be TASMOTA-Last6MacNumbers-Last4MacNumbers

If you set Topic: `test`, the Hostname will be `test-7318`

If you use %variables% inside a name, they WON'T WORK unless used inside a rule. Using %variables% in the console are not supported. Variables are only for RULES.

So, for what you are trying to achieve, one option is to define a hostname and a topic in your script. Another option is to self-compile and add rules that will do those changes at first boot to the parameters you want. An example of that is:
In **user_config_override.h** add:
```cpp
#define USER_RULE1            "ON power1#boot DO backlog0 hostname AWESOME-%macaddr%;rule1 0;rule1 \" ENDON"
#define USER_BACKLOG          "Rule1 1;topic %12X"
```
The previous example runs only after the first boot OR after a `reset 1` command. What it does is to set HOSTNAME to AWESOME-MACaddress and set the TOPIC as MACaddress. Topic and Hostname are now independent with this. And after the execution of all commands, the rule is cleared.

Just for the sake of leaving here the complete information for anyone interested, following are the restrictions for HOSTNAME (also written in the Tasmota docs):

> Hostname
> 
> Should not exceed 14 characters. Valid characters include alphanumerical (A–Z, a–z, 0–9), and the hyphen (-). The first character must be a letter. You cannot use "localhost" as hostname for a device.
> 
> Note
> Max is 60 characters but many routers and NetBios systems don't support that. Stick to 14 or less for maximum compatibility.
> It is recommended that you use lowercase letters to ensure that certificate authentication in Cisco ISE is not impacted by minor differences in certificate-driven verifications.

Hope this helps to understand the behavior of Tasmota and configuration options.
